File: dir_m_dependencies/parameter_timing_update.m
function parameter = parameter_timing_update(parameter,str_field,dt,ncall,nop);
if (nargin<3); dt=0; end;
if (nargin<4); ncall=1; end;
if (nargin<5); nop=0; end;

if isempty(parameter);
parameter = struct('type','parameter');
end;%if isempty(parameter);
%%%%%%%%;
if (~isfield(parameter,'timing'));
parameter.timing = cell(1,6);
parameter.timing{1,1} = 'label';
parameter.timing{1,2} = 'total time';
parameter.timing{1,3} = 'number of calls';
parameter.timing{1,4} = 'number of ops';
parameter.timing{1,5} = 'MHz';
parameter.timing{1,6} = 'GHz';
end; %<-- parameter_bookmark. ;
%%%%%%%%;
tmp_index = min(efind(strcmp(parameter.timing(:,1),str_field)));
if (~isempty(tmp_index));
parameter.timing{1+tmp_index,2} = parameter.timing{1+tmp_index,2}+dt;
parameter.timing{1+tmp_index,3} = parameter.timing{1+tmp_index,3}+ncall;
parameter.timing{1+tmp_index,4} = parameter.timing{1+tmp_index,4}+nop;
end;%if ( isempty(tmp_index));
if ( isempty(tmp_index));
tmp_index = size(parameter.timing,1);
parameter.timing{1+tmp_index,1} = str_field;
parameter.timing{1+tmp_index,2} = dt;
parameter.timing{1+tmp_index,3} = ncall;
parameter.timing{1+tmp_index,4} = nop;
end;%if ( isempty(tmp_index));
%%%%%%%%;
n_op = parameter.timing{1+tmp_index,4};
d_t = parameter.timing{1+tmp_index,2};
if ( (n_op>0) & (d_t>0) );
parameter.timing{1+tmp_index,5} = (n_op/d_t)/1e6;
parameter.timing{1+tmp_index,6} = (n_op/d_t)/1e9;
end;%if ( (n_op>0) & (d_t>0) );
%%%%%%%%;

File: dir_rangan_python/dir_matlab_macros/parameter_timing_update.py
#import os; os.chdir('/data/rangan/dir_cryoem/dir_rangan_python');
import numpy as np ;
from disp_sprintf import disp ; from disp_sprintf import sprintf ;

def parameter_timing_update(
        parameter=None,
        str_field=None,
        dt=None,
        ncall=None,
        nop=None,
):
    if dt is None: dt=0.0;
    if ncall is None: ncall=int(0);
    if nop is None: nop=int(0);
    if parameter is None: parameter = {'type': 'parameter'};
    if 'timing' not in parameter or parameter['timing'] is None:
        parameter['timing'] = [
            ['label', 'total time', 'number of calls', 'number of ops', 'MHz', 'GHz']
        ];
    #end;%if ;
    timing = parameter['timing'];
    tmp_index = None;
    for idx in range(1, len(timing)):
        if timing[idx][0] == str_field:
            tmp_index = idx ;
            break ;
        #end;%if;
    #end;%for;

    if tmp_index is not None:
        timing[tmp_index][1] = timing[tmp_index][1] + dt ;
        timing[tmp_index][2] = timing[tmp_index][2] + ncall ;
        timing[tmp_index][3] = timing[tmp_index][3] + nop ;
    else:
        timing.append([str_field, dt, ncall, nop, 0.0, 0.0]) ;
        tmp_index = len(timing) - 1 ;
    #end;%if;

    n_op = timing[tmp_index][3] ;
    d_t = timing[tmp_index][1] ;
    if (n_op > 0) and (d_t > 0):
        timing[tmp_index][4] = (n_op / np.maximum(1e-12,d_t)) / 1e6 ; # MHz
        timing[tmp_index][5] = (n_op / np.maximum(1e-12,d_t)) / 1e9 ; # GHz
    #end;%if;

    parameter['timing'] = timing;
    return(parameter);


